src/node_report.cc

```diff
@@ -0,0 +1,1024 @@
+  while (env_var != nullptr) {
+    pair = std::string(env_var);
+    int separator = pair.find('=');
+    json_keyvaluenormal(
+        out, *indent, pair.substr(0, separator), pair.substr(separator + 1));
+    env_var = *(environ + index++);
+  }
+  unread(out);
+  json_deindent(indent);
+  json_objectendnormal(out, *indent);
+
+  json_objectstart(out, *indent, "user_limits");
+  json_indent(indent);
+  struct rlimit limit;
+  char buf[64];
+  std::string soft, hard;
+
+  for (size_t i = 0; i < arraysize(rlimit_strings); i++) {
+    if (getrlimit(rlimit_strings[i].id, &limit) == 0) {
+      if (limit.rlim_cur == RLIM_INFINITY) {
+        soft = std::string("unlimited");
+      } else {
+#if defined(_AIX) || defined(__sun)
+        snprintf(buf, sizeof(buf), "%ld", limit.rlim_cur);
+        soft = std::string(buf);
+#elif defined(__linux__) && !defined(__GLIBC__)
+        snprintf(buf, sizeof(buf), "%ld", limit.rlim_cur);
+        soft = std::string(buf);
+#else
+        snprintf(buf, sizeof(buf), "%16" PRIu64, limit.rlim_cur);
+        soft = std::string(soft);
+#endif
+      }
+      if (limit.rlim_max == RLIM_INFINITY) {
+        hard = std::string("unlimited");
+      } else {
+#if defined(_AIX)
+        snprintf(buf, sizeof(buf), "%lu", limit.rlim_max);
+        hard = std::string(buf);
+#elif(defined(__linux__) && !defined(__GLIBC__))
+        snprintf(buf, sizeof(buf), "%lu", limit.rlim_max);
+        hard = std::string(buf);
+#else
+        snprintf(buf, sizeof(buf), "%lu", limit.rlim_max);
+        hard = std::string(buf);
+#endif
+      }
+      json_objectstart(out, *indent, rlimit_strings[i].description);
+      json_indent(indent);
+      json_keyvaluenormal(out, *indent, "soft", soft);
+      json_keyvaluefinal(out, *indent, "hard", hard);
+      json_deindent(indent);
+      json_objectendnormal(out, *indent);
+    }
+  }
+  unread(out);
+  json_deindent(indent);
+  json_objectendnormal(out, *indent);
+#endif
+
+  PrintLoadedLibraries(out, indent);
+}
+
+/*******************************************************************************
+ * Functions to print a list of loaded native libraries.
+ *
+ ******************************************************************************/
+#ifdef __linux__
+static int LibraryPrintCallback(struct dl_phdr_info* info,
+                                size_t size, void* data) {
+  // std::ostream* out = static_cast<std::ostream*>(data);
+  write_data* wrap = reinterpret_cast<write_data*>(data);
+  std::ostream* out = wrap->out;
+  int* indent = wrap->indent;
+  if (info->dlpi_name != nullptr && *info->dlpi_name != '\0') {
+    json_arrayentrynormal(*out, *indent, info->dlpi_name);
+  }
+  return 0;
+}
+#endif
+
+static void PrintLoadedLibraries(std::ostream& out, int* indent) {
+#ifdef __linux__
+  write_data data;
+  data.out = &out;
+  data.indent = indent;
+  json_arraystart(out, *indent, "shared_objects");
+  json_indent(indent);
+  dl_iterate_phdr(LibraryPrintCallback, &data);
```

Comment:
I think this would be a case where it’s okay to make `LibraryPrintCallback` a lambda function? That way the code is written in the order in which it executes, which can help with readability

Reply:
thanks, converted to a lambda.I wanted to optimize it by introducing the JSONWriter pointer as a `catpure` to the lambda as:
```c++
  dl_iterate_phdr([&writer](struct dl_phdr_info* info, size_t size, void* data) {
    if (info->dlpi_name != nullptr && *info->dlpi_name != '\0') {
      writer->json_element(info->dlpi_name);
    }
    return 0;
  }, nullptr);
```
however the compiler throws cryptic error:
```error
../src/node_report.cc: In function ‘void nodereport::PrintLoadedLibraries(nodereport::JSONWriter*)’:
../src/node_report.cc:799:13: error: cannot convert ‘nodereport::PrintLoadedLibraries(nodereport::JSONWriter*)::__lambda6’ to ‘int (*)(dl_phdr_info*, size_t, void*) {aka int (*)(dl_phdr_info*, long unsigned int, void*)}’ for argument ‘1’ to ‘int dl_iterate_phdr(int (*)(dl_phdr_info*, size_t, void*), void*)’
   }, nullptr);
```

can you please advise what is happening here and how do we fix it? I am new to c++ lambda.